R/data.table.R

```diff
@@ -481,8 +481,8 @@ replace_dot_alias = function(e) {
           allow.cartesian = TRUE
         }
         # TODO: collect all '==' ops first to speeden up Cnestedid
-        rightcols = colnamesInt(x, names(on), check_dups=FALSE)
-        leftcols  = colnamesInt(i, unname(on), check_dups=FALSE)
+        rightcols = colnamesInt(x, names(on), check_dups=FALSE, source = "data.table.R(ln:484)" )
+        leftcols  = colnamesInt(i, unname(on), check_dups=FALSE, source = "data.table.R(ln:484)")
```

Comment:
Thanks for getting this started! As mentioned, knowing the data.table source file/line is not very useful for an end user, only for data.table developers. If you are running analysis and encounter an error, the last thing you want to do is have to navigate to GitHub and find where this message comes from. Moreover, pointing to the source code will be useless to the vast majority of users, who don't want to wade through implementation details to figure out what's gone wrong. Lastly, a simple stack trace from `traceback()` should provide as much information as this message (ideally, though our source is currently a bit of a mess / harder to navigate from a traceback than it ideally should be.

In short, the thing we really want to focus on as our north star when trying to improve error messages is the end user, be they beginner/intermediate/advanced in their fluency with data.table.

With that in mind, what I was thinking of was something like

```suggestion
        rightcols = colnamesInt(x, names(on), check_dups=FALSE, source = "matching columns from 'on' names to columns in 'x'")
        leftcols  = colnamesInt(i, unname(on), check_dups=FALSE, source = "matching column names from 'on' values to columns in 'i'")
```

Then we can go from an error like


```r
data.table:::colnamesInt(list(a = 1), 'b')
# Error in data.table:::colnamesInt(list(a = 1, b = 2), "c") : 
#   argument specifying columns received non-existing column(s): cols[1]='c'

# example improvement
#   In matching columns from 'on' names to columns in 'x', received non-existing column(s): cols[1]='c'
```

Even this doesn't feel perfect, but it's in the right direction. Hopefully that clarifies the goal of this FR & how to focus your efforts in the PR.